File: 백준/Gold/17144. 미세먼지 안녕！/미세먼지 안녕！.py
'''
미세먼지 확산 / 모든 칸 동시에 / 인접한 4방향 / r,c/5만큼 소수점 버림 /남은 것 rc - rc/5* 확산 방향 수
공기청정기 작동/ 바람 나옴 / 위쪽 : 반시계 / 아래쪽 시계 방향/ 미세먼지가 바람 방향대로 한칸 이동/ 공기청정기로 들어가면 정화됨
'''
import sys

input = sys.stdin.readline


def dust():
    arr2 = [[0] * c for _ in range(r)]

    for i in range(r):
        for j in range(c):
            if arr[i][j] >= 5:
                tmp = arr[i][j] // 5
                for d in range(4):
                    nx = i + dx1[d]
                    ny = j + dy1[d]
                    if 0 <= nx < r and 0 <= ny < c and not exist[nx][ny]:  # 공기 청정기 칸은 확산일어나지 않음
                        arr2[nx][ny] += tmp  # arr2에 늘어나는 미세먼지양을 저장함
                        arr2[i][j] -= tmp  # arr2에 감소하는 양을 빼줌

    for i in range(r):
        for j in range(c):  # arr에 arr2값 더해주기
            arr[i][j] += arr2[i][j]


def clear_dust():
    # 위
    d = 0
    x, y = clear[0][0], 1
    tmp = 0
    while True:
        nx = x + dx1[d]
        ny = y + dy1[d]

        if x == clear[0][0] and y == 0:
            break
        if (nx < 0 or nx >= r) or (ny < 0 or ny >= c):
            d += 1
            continue

        arr[x][y], tmp = tmp, arr[x][y]
        x, y = nx, ny
    # 아래
    d = 0
    x, y = clear[1][0], 1
    tmp = 0
    while True:
        nx = x + dx2[d]
        ny = y + dy2[d]

        if x == clear[1][0] and y == 0:
            break
        if (nx < 0 or nx >= r) or (ny < 0 or ny >= c):
            d += 1
            continue

        arr[x][y], tmp = tmp, arr[x][y]
        x, y = nx, ny


r, c, t = map(int, input().split())
exist = [[False] * c for _ in range(r)]
clear = []
arr = []

for i in range(r):
    tmp = list(map(int, input().split()))
    for j in range(c):
        if tmp[j] == -1:
            tmp[j] = 0
            exist[i][j] = True
            clear.append([i, j])
    arr.append(tmp)

#
dx1 = [0, -1, 0, 1]
dy1 = [1, 0, -1, 0]

dx2 = [0, 1, 0, -1]
dy2 = [1, 0, -1, 0]
for _ in range(t):
    dust()
    # print()
    # print(*arr, sep='\n')
    clear_dust()
    # print()
    # print(*arr, sep='\n')

print(sum(sum(arr, [])))

File: 백준/Gold/17144. 미세먼지 안녕！/미세먼지 안녕！.java
import java.io.*;
import java.util.*;

public class Main {
	static int r, c, t;
	static boolean[][] exist;
	static int[][] arr;
	static int[] dx1 = {0, -1, 0, 1};
	static int[] dy1 = {1, 0, -1, 0};
	static int[] dx2 = {0, 1, 0, -1};
	static int[] dy2 = {1, 0, -1, 0};
	static List<int[]> clear;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        t = Integer.parseInt(st.nextToken());

        exist = new boolean[r][c];
        clear = new ArrayList<>();
        arr = new int[r][c];

        for (int i = 0; i < r; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < c; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
                if (arr[i][j] == -1) {
                    arr[i][j] = 0;
                    exist[i][j] = true;
                    clear.add(new int[]{i, j});
                }
            }
        }

        

        for (int time = 0; time < t; time++) {
            dust();
            clearDust();
        }

        int sum = 0;
        for (int[] row : arr) {
            for (int val : row) {
                sum += val;
            }
        }

        System.out.println(sum);
    }

    private static void dust() {
        int[][] arr2 = new int[r][c];

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (arr[i][j] >= 5) {
                    int tmp = arr[i][j] / 5;
                    for (int d = 0; d < 4; d++) {
                        int nx = i + dx1[d];
                        int ny = j + dy1[d];
                        if (0 <= nx && nx < r && 0 <= ny && ny < c && !exist[nx][ny]) {
                            arr2[nx][ny] += tmp;
                            arr2[i][j] -= tmp;
                        }
                    }
                }
            }
        }

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                arr[i][j] += arr2[i][j];
            }
        }
    }

    private static void clearDust() {
        int d = 0;
        int x = clear.get(0)[0];
        int y = 1;
        int tmp = 0;
        while (true) {
            int nx = x + dx1[d];
            int ny = y + dy1[d];

            if (x == clear.get(0)[0] && y == 0) {
                break;
            }
            if (nx < 0 || nx >= r || ny < 0 || ny >= c) {
                d++;
                continue;
            }

            int temp = arr[x][y];
            arr[x][y] = tmp;
            tmp = temp;
            x = nx;
            y = ny;
        }


        d = 0;
        x = clear.get(1)[0];
        y = 1;
        tmp = 0;
        while (true) {
            int nx = x + dx2[d];
            int ny = y + dy2[d];

            if (x == clear.get(1)[0] && y == 0) {
                break;
            }
            if (nx < 0 || nx >= r || ny < 0 || ny >= c) {
                d++;
                continue;
            }

            int temp = arr[x][y];
            arr[x][y] = tmp;
            tmp = temp;
            x = nx;
            y = ny;
        }
    }
}
